coderd/database/gen/metrics/main.go

```diff
@@ -0,0 +1,210 @@
+package main
+
+import (
+	"fmt"
+	"go/format"
+	"go/token"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/dave/dst"
+	"github.com/dave/dst/decorator"
+	"github.com/dave/dst/decorator/resolver/goast"
+	"github.com/dave/dst/decorator/resolver/guess"
+	"golang.org/x/xerrors"
+)
+
+func main() {
+	err := run()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run() error {
+	funcs, err := readStoreInterface()
+	if err != nil {
+		return err
+	}
+	funcByName := map[string]struct{}{}
+	for _, f := range funcs {
+		funcByName[f.Name] = struct{}{}
+	}
+	declByName := map[string]*dst.FuncDecl{}
+
+	dbmetrics, err := os.ReadFile("./dbmetrics/dbmetrics.go")
+	if err != nil {
+		return xerrors.Errorf("read dbfake: %w", err)
+	}
+
+	// Required to preserve imports!
+	f, err := decorator.NewDecoratorWithImports(token.NewFileSet(), "dbmetrics", goast.New()).Parse(dbmetrics)
+	if err != nil {
+		return xerrors.Errorf("parse dbfake: %w", err)
+	}
+
+	for i := 0; i < len(f.Decls); i++ {
+		funcDecl, ok := f.Decls[i].(*dst.FuncDecl)
+		if !ok || funcDecl.Recv == nil || len(funcDecl.Recv.List) == 0 {
+			continue
+		}
+		// Check if the receiver is the struct we're interested in
+		_, ok = funcDecl.Recv.List[0].Type.(*dst.Ident)
+		if !ok {
+			continue
+		}
+		if _, ok := funcByName[funcDecl.Name.Name]; !ok {
+			continue
+		}
+		declByName[funcDecl.Name.Name] = funcDecl
+		f.Decls = append(f.Decls[:i], f.Decls[i+1:]...)
+		i--
+	}
+
+	for _, fn := range funcs {
+		decl, ok := declByName[fn.Name]
+		if !ok {
+			params := make([]string, 0)
+			if fn.Func.Params != nil {
+				for _, p := range fn.Func.Params.List {
+					for _, name := range p.Names {
+						params = append(params, name.Name)
+					}
+				}
+			}
+			returns := make([]string, 0)
+			if fn.Func.Results != nil {
+				for i := range fn.Func.Results.List {
+					returns = append(returns, fmt.Sprintf("r%d", i))
+				}
+			}
+
+			code := fmt.Sprintf(`
+package stub
+
+func stub() {
+	start := time.Now()
+	%s := m.s.%s(%s)
+	m.queryLatencies.WithLabelValues("%s").Observe(time.Since(start).Seconds())
+	return %s
+}
+`, strings.Join(returns, ","), fn.Name, strings.Join(params, ","), fn.Name, strings.Join(returns, ","))
```

Comment:
Suggestion, but up to you. The dave package said it retained comments, but not 100% if this works.

```golang
package stub
func stub() {
        // Function generated by coderd/database/gen/metrics/main.go
        // Manual edits are allowed and will be retained.
	start := time.Now()
	%s := m.s.%s(%s)
	m.queryLatencies.WithLabelValues("%s").Observe(time.Since(start).Seconds())
	return %s
}
````

The general idea is just to inform the developers where this code comes from. Especially if we run into any issues early on.

Your call though.

Reply:
That would work, but I think I'll exclude it for now... I don't make note that it's generated anyways, so I think people will feel free to edit it